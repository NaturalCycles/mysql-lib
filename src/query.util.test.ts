import { DBQuery } from '@naturalcycles/db-lib'
import {
  createTestItemDBM,
  createTestItemsDBM,
  TEST_TABLE,
} from '@naturalcycles/db-lib/dist/testing'
import { _range } from '@naturalcycles/js-lib'
import { dbQueryToSQLDelete, dbQueryToSQLSelect, dbQueryToSQLUpdate, insertSQL } from './query.util'

test('dbQueryToSQLSelect', () => {
  let sql = dbQueryToSQLSelect(new DBQuery('TBL1'))
  expect(sql).toMatchSnapshot()

  sql = dbQueryToSQLSelect(
    new DBQuery('TBL1')
      .filterEq('a', 'b')
      .filter('c', '>', '2019')
      .order('aaa')
      .order('bbb', true)
      .limit(15),
  )
  // console.log(sql)
  expect(sql).toMatchSnapshot()

  sql = dbQueryToSQLSelect(new DBQuery('TBL1').filter('num', '>', 15))
  // console.log(sql)
  expect(sql).toMatchSnapshot()

  // NULL cases
  sql = dbQueryToSQLSelect(
    new DBQuery('TBL1').filter('a', '=', undefined).filter('a2', '=', null).filter('a3', '>', null),
  )
  // console.log(sql)
  expect(sql).toMatchSnapshot()

  // ARRAY CASES
  sql = dbQueryToSQLSelect(new DBQuery('TBL1').filter('a', '=', ['a1', 'a2', 'a3']))
  // console.log(sql)
  expect(sql).toMatchSnapshot()
})

test('dbQueryToSQLDelete', () => {
  let sql = dbQueryToSQLDelete(new DBQuery('TBL1'))
  expect(sql).toMatchSnapshot()

  sql = dbQueryToSQLDelete(new DBQuery('TBL1').filter('a', '>', null))
  expect(sql).toMatchSnapshot()
})

test('insertSQL', () => {
  const items = createTestItemsDBM(3)
  const [sql] = insertSQL(TEST_TABLE, items)
  // console.log(sql)
  expect(sql).toMatchSnapshot()
})

test('dbQueryToSQLUpdate', () => {
  const item = createTestItemDBM()

  let sql = dbQueryToSQLUpdate(new DBQuery(TEST_TABLE), item)
  // console.log(sql)
  expect(sql).toMatchSnapshot()

  sql = dbQueryToSQLUpdate(new DBQuery(TEST_TABLE).filter('a', '>', 5), item)
  // console.log(sql)
  expect(sql).toMatchSnapshot()
})

test('large sql query split', () => {
  const items = createTestItemsDBM(10).map(r => ({ ...r, lng: 'xxx'.repeat(80000) }))
  const sqls = insertSQL(TEST_TABLE, items)
  // console.log(sqls)
  console.log(sqls.length)
  expect(sqls.length).toBeGreaterThan(1)
})
