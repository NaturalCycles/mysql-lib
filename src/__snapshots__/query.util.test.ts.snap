// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbQueryToSQLDelete 1`] = `"DELETE FROM \`TBL1\`"`;

exports[`dbQueryToSQLDelete 2`] = `"DELETE FROM \`TBL1\` WHERE \`a\` IS NOT NULL"`;

exports[`dbQueryToSQLSelect 1`] = `"SELECT * FROM \`TBL1\`"`;

exports[`dbQueryToSQLSelect 2`] = `"SELECT * FROM \`TBL1\` WHERE \`a\` = 'b' AND \`c\` > '2019' ORDER BY \`aaa\` ASC, \`bbb\` DESC LIMIT 15"`;

exports[`dbQueryToSQLSelect 3`] = `"SELECT * FROM \`TBL1\` WHERE \`num\` > 15"`;

exports[`dbQueryToSQLSelect 4`] = `"SELECT * FROM \`TBL1\` WHERE \`a\` IS NULL AND \`a2\` IS NULL AND \`a3\` IS NOT NULL"`;

exports[`dbQueryToSQLSelect 5`] = `"SELECT * FROM \`TBL1\` WHERE \`a\` IN ('a1', 'a2', 'a3')"`;

exports[`dbQueryToSQLUpdate 1`] = `"UPDATE \`TEST_TABLE\` SET \`id\` = 'id1', \`k1\` = 'v1', \`k2\` = 'v2', \`k3\` = 1, \`even\` = false, \`created\` = 1529539200, \`updated\` = 1529539200"`;

exports[`dbQueryToSQLUpdate 2`] = `"UPDATE \`TEST_TABLE\` SET \`id\` = 'id1', \`k1\` = 'v1', \`k2\` = 'v2', \`k3\` = 1, \`even\` = false, \`created\` = 1529539200, \`updated\` = 1529539200 WHERE \`a\` > 5"`;

exports[`insertSQL 1`] = `
"INSERT INTO \`TEST_TABLE\` (\`id\`,\`k1\`,\`k2\`,\`k3\`,\`even\`,\`created\`,\`updated\`) VALUES
('id1','v1','v2',1,false,1529539200,1529539200),
('id2','v2','v4',2,true,1529539200,1529539200),
('id3','v3','v6',3,false,1529539200,1529539200)"
`;
